lib/pushmi_pullyu/preservation_queue.rb

```diff
@@ -0,0 +1,79 @@
+require 'redis'
+require 'connection_pool'
+
+# Quick and dirty take on sorted sets
+
+# 1) Create a sorted set in Redis (https://redis.io/topics/data-types). Call it preservation_queue
+#
+# 2) In GenericFile add an after_save that:
+#    - determines a monotonically increasing "score". Obvious scores would be either the time in seconds/milliseconds
+#      or using something like redis INCR to create an atomic, increasing counter. It doesn't matter if 2 different
+#      noids ever have the same score, it only that scores generally increase over time.
+#    - zadd preservation_queue score "noid" adds the noid and gives it the score from above.
+#
+# 3) Pushmi-pullyu pops elements out of the sorted set, lowest score to highest.
+#
+# A sorted set will only ever contain a noid once, with whatever score it was last given. Because preservation_queue
+# is sorted lowest score to highest, and because scores increase over time, a cascade of jobs/updates will cause a noid
+# to keep "moving back" in the queue until it becomes the least recently updated noid in the queue, at which point it
+# will be popped and preserved. Any further updates will trigger a new AIP build.
+class PushmiPullyu::PreservationQueue
+
+  class ConnectionError < StandardError; end
+
+  def initialize(connection: {}, pool_opts: { size: 1, timeout: 5 }, poll_interval: 10, age_at_least: 0,
```

Comment:
I didn't bother to add everything, but I added CLI option flags for the ones we're most likely to need to change immediately.